robosat/tools/train.py

```diff
@@ -73,39 +75,75 @@ def main(args):
     if model["common"]["cuda"]:
         torch.backends.cudnn.benchmark = True
 
+    weight = torch.Tensor(dataset["weights"]["values"])
+    optimizer = Adam(net.parameters(), lr=model["opt"]["lr"], weight_decay=model["opt"]["decay"])
+
+    resume = 0
     if args.checkpoint:
 
         def map_location(storage, _):
             return storage.cuda() if model["common"]["cuda"] else storage.cpu()
 
         # https://github.com/pytorch/pytorch/issues/7178
         chkpt = torch.load(args.checkpoint, map_location=map_location)
-        net.load_state_dict(chkpt)
-
-    optimizer = Adam(net.parameters(), lr=model["opt"]["lr"], weight_decay=model["opt"]["decay"])
-
-    weight = torch.Tensor(dataset["weights"]["values"])
-
-    criterion = CrossEntropyLoss2d(weight=weight).to(device)
-    # criterion = FocalLoss2d(weight=weight).to(device)
+        net.load_state_dict(chkpt["state_dict"])
+
+        if args.resume:
+            optimizer.load_state_dict(chkpt["optimizer"])
+            resume = chkpt["epoch"]
+
+    if model["opt"]["loss"] == "CrossEntropy":
+        criterion = CrossEntropyLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "mIoU":
+        criterion = mIoULoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Focal":
+        criterion = FocalLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Lovasz":
+        criterion = LovaszLoss2d().to(device)
+    else:
+        sys.exit("Error: Unknown [opt][loss] value !")
 
     train_loader, val_loader = get_dataset_loaders(model, dataset, args.workers)
 
     num_epochs = model["opt"]["epochs"]
+    if resume >= num_epochs:
+        sys.exit("Error: Epoch {} set in {} already reached by the checkpoint provided".format(num_epochs, args.model))
 
     history = collections.defaultdict(list)
+    log = Log(os.path.join(model["common"]["checkpoint"], "log"))
+
+    log.log("--- Hyper Parameters on Dataset: {} ---".format(dataset["common"]["dataset"]))
+    log.log("Batch Size:\t {}".format(model["common"]["batch_size"]))
+    log.log("Image Size:\t {}".format(model["common"]["image_size"]))
+    log.log("Learning Rate:\t {}".format(model["opt"]["lr"]))
+    log.log("LR Decay:\t {}".format(model["opt"]["decay"]))
```

Comment:
This is not the learning rate decay but instead the l2 regularization decay for Adam; not sure we should use it in the first place since what we really want is AdamW and not the l2 weight decay implemented in PyTorch currently.

Reply:
Right now, the only losses prototypes are with or without weights.
But we could imagine adding per batch rather than per image.
Or if Focal is useful adding gamma in userland.
And so on...

So i'm not convinced at this point, that in the end, it would be more elegant/simpler 
to read and maintain...



Thanks for Adam's Weight Decay. And i updated the label accordingly.

Reply:
:+1: